### PR TITLE
dx: fix stack traces for device.*() calls

### PR DESCRIPTION
**Pull request by noomorph**
## Description

This is a developer experience improvement.

It improves stack traces for known errors (derived from Detox) occurring inside asynchronous `device.***()` public methods, e.g.:

#### Before the fix

device.launchApp() stack trace before fix

#### After the fix

device.launchApp() stack trace after fix

As you can see, the improved error stack points exactly to the user code, and not to the misleading lines in Detox internals.

## Test strategy

Added an `e2e-unhappy` suite for this case. Regression tests on CI should pass as well before this can be merged.

### CODE REVIEW

**Review by d4vidi — COMMENTED** [review]
@noomorph 
1. Kudos for keeping the DevEx our top prio!
2. Please review some inline notes...
3. Question: Why was the `device` API "special" so as to require this tailored handling, in the first place?

File: detox/src/devices/Device.js [hunk 1]
```diff
@@ -16,6 +17,41 @@ class Device {
...
+      'setURLBlacklist',
+      'shake',
+      'shutdown',
+      'takeScreenshot',
+      'terminateApp',
+      'uninstallApp',
+      'unmatchFace',
+      'unmatchFinger',
+      'unreverseTcpPort',
+    ]);
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/devices/Device.js:53]: I'm running only through public methods we ± documented, and all of them are async. Need to think about that, indeed.

File: detox/src/utils/cutStackTraces.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+const cutStackTraces = require('./cutStackTraces');
+
+describe('cutStackTraces', () => {
+  it.each([
+    ['DetoxError'],
+    ['DetoxConfigError'],
+    ['DetoxInternalError'],
+    ['DetoxRuntimeError'],
+  ])('should post-process known errors (e.g., %s)', async (errName) => {
+    const AnError = errors[errName];
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/utils/cutStackTraces.test.js:12]: Will do, thanks.

File: detox/src/utils/cutStackTraces.test.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    const obj = createThrowingObject(AnError);
+
+    /** @type {Error} */
+    const anError = await obj.willThrow().catch(e => e);
+    expect(anError).toBeInstanceOf(AnError);
+    expect(anError.stack).not.toBe(obj.originalErrorStack);
+    expect(anError.stack).not.toContain('detox/src');
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/utils/cutStackTraces.test.js:19]: Ok, I'll add a check for `__filename` to be inside the stack trace.

File: detox/src/utils/cutStackTraces.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+const DetoxError = require('../errors/DetoxError');
+const { asError, createErrorWithUserStack, replaceErrorStack } = require('../utils/errorUtils');
+
+function cutStackTraces(obj, methodNames) {
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/utils/cutStackTraces.js:4]: `wrapWithStackTraceCutter` maybe?

[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/src/utils/cutStackTraces.js:4]: Heh go wild

File: detox/src/devices/Device.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/src/devices/Device.js:53]: We might try some AOP if we add TypeScript btw. 🤔 
image

**Pull request merged**